### PR TITLE
Update scripts using UTF-8

### PR DESCRIPTION
**Pull request by nattofriends**
Read and write script files using UTF-8 instead of ascii.

### CODE REVIEW

**Comment by asottile** [comment]
Could you write a quick test which demonstrates the change? I have a feeling this actually breaks python 2 which was handling the file entirely with bytes.

Also please use `io.open` instead of codecs as it has more familiar python 3 semantics.

Do we care if a file is non UTF8? Should we be parsing entirely as bytes and be encoding agnostic?

**Comment by nattofriends** [comment]
Yeah, Python 2 was actually just fine treating the file as bytes. I finally nailed down my actual problem, which related to the fact that the encoding of files is determined from `locale.getpreferredencoding()`, which defaults to `ascii` in the absence of the environment variable `LANG`.

I tried your suggestion to parse as bytes only and it seems to have worked in both Python 2 and 3, I'll post that along with actual tests in a bit.

File: tests/virtualenv_tools_test.py [hunk 1]
```diff
@@ -13,27 +13,41 @@ def auto_namedtuple(**kwargs):
     return collections.namedtuple('ns', tuple(kwargs))(**kwargs)
 
 
-@pytest.yield_fixture
-def venv(tmpdir):
+@pytest.yield_fixture(params=[None, '-e'])
```
Comment (asottile) [inline tests/virtualenv_tools_test.py]: personally dislike `params` as I find it hard to read -- also this doubles the length of the test run

Please just install two modules instead -- it's more explicit and doesn't double the testsuite size

File: tests/virtualenv_tools_test.py [hunk 2]
```diff
@@ -13,27 +13,41 @@ def auto_namedtuple(**kwargs):
...
+def venv(tmpdir, request):
     app_before = tmpdir.join('before').ensure_dir()
     app_before.join('mymodule.py').write(
```
Comment (asottile) [inline tests/virtualenv_tools_test.py]: there's a `write_binary`

File: tests/virtualenv_tools_test.py [hunk 3]
```diff
@@ -13,27 +13,41 @@ def auto_namedtuple(**kwargs):
...
+        b"    print('ohai!')\n",
+        mode='wb'
     )
     app_before.join('setup.py').write(
         'from setuptools import setup\n'
-        'setup(name="mymodule", py_modules=["mymodule"])\n'
+        'setup('
+        'name="mymodule", '
+        'py_modules=["mymodule"], '
+        'scripts=["mymodule.py"])\n'
```
Comment (asottile) [inline tests/virtualenv_tools_test.py]: p2: format this nicely so it's readable code

```python
'setup(\n'
'    name="mymodule",\n'
...
```

**Review by asottile — CHANGES_REQUESTED** [review]

File: tests/virtualenv_tools_test.py (hunk 1 above)
Comment (nattofriends) [reply tests/virtualenv_tools_test.py]: The reason for installing non-editable is to test the rewriting of scripts copied verbatim into venv/bin/, and installing editable to test the rewriting of pth files. The editable flag affects everything referenced in the setup.py so I'm not sure how making two modules (mymodule1 and mymodule2, if I understand you correctly) would help?

Comment (asottile) [reply tests/virtualenv_tools_test.py]: install one module editable, one module not

Comment (asottile) [reply tests/virtualenv_tools_test.py]: or better yet, make a single test which installs a non-editable module into `venv`

File: virtualenv_tools.py [hunk 4]
```diff
@@ -91,9 +94,9 @@ def update_script(script_filename, old_path, new_path):
         return
 
     args[0] = new_bin
-    lines[0] = '#!%s\n' % ' '.join(args)
+    lines[0] = b'#!%s\n' % b' '.join(args)
```
Comment (asottile) [inline virtualenv_tools.py]: in python3 this uses a python3.5+ feature, 3.4 isn't end-of-lifed yet so maybe we should use more compatible syntax?  Though yelp doesn't have any py34 left does it?

File: tests/virtualenv_tools_test.py [hunk 5]
```diff
@@ -111,6 +111,49 @@ def test_move(venv, capsys):
...
+    subprocess.check_call(cmd)
+    subprocess.check_call((
+        venv_before.join('bin/pip').strpath,
+        'install', app_before.strpath,
+    ))
+
+    venv = auto_namedtuple(
+        app_before=app_before, app_after=app_after,
+        before=venv_before, after=venv_after,
+    )
```
Comment (asottile) [inline tests/virtualenv_tools_test.py]: no need to make this namedtuple, the structure isn't used after this

File: tests/virtualenv_tools_test.py [hunk 6]
```diff
@@ -111,6 +111,49 @@ def test_move(venv, capsys):
     assert_virtualenv_state(venv.after)
 
 
+def test_move_installed_not_editable(tmpdir, capsys):
```
Comment (asottile) [inline tests/virtualenv_tools_test.py]: why not reuse the `venv` fixture and just install into it?

Comment (asottile) [inline tests/virtualenv_tools_test.py]: this test name should probably indicate what it's regression testing: a non-utf8 script in bin

**Review by asottile — COMMENTED** [review]
otherwise lg2m

File: tests/virtualenv_tools_test.py (hunk 6 above)
Comment (nattofriends) [reply tests/virtualenv_tools_test.py]: The script itself is utf8, but py3 will try to read it as ascii if LANG isn't defined.

Comment (asottile) [reply tests/virtualenv_tools_test.py]: err yes, I meant non-ascii, good catch :)

**Review by asottile — APPROVED** [review]

**Pull request merged**